Unit tests fails for fresh generated generator

```
$ yarn mocha dist/tests/**/*.js
$ /tmp/testing/generate-generator-example/node_modules/.bin/mocha 'dist/tests/**/*.js'


  testing
    tasks
      1) should extend tasks onto the instance
      ✓ should run the `default` task with .build (167ms)
      ✓ should run the `default` task with .generate
    generator-example (CLI)
      2) should run the default task using the `testing` name (global install)
      3) should run the default task using the `generator-example` generator alias (local generator.js)
    generator-example (API)
      ✓ should run the default task on the generator
Cannot find task: "project" in generator: "generator-example"
      4) should run the `generator-example` task
      ✓ should run the `default` task when defined explicitly
    sub-generator
      ✓ should work as a sub-generator
      ✓ should run the `default` task by default
      ✓ should run the `generator:default` task when defined explicitly
Cannot find task: "project" in generator: "foo.generator-example"
      5) should run the `generator:project` task when defined explicitly
      ✓ should work with nested sub-generators
Cannot find task: "package" in generator: "sub0.sub1"
      6) should run tasks as a sub-generator

  testing
    plugin
      ✓ should add tasks to the instance
      ✓ should only register the plugin once


  10 passing (4s)
  6 failing

  1) testing
       tasks
         should extend tasks onto the instance:
     AssertionError: expected { Object (check-directory, main, ...) } to have property 'project'
      at Context.<anonymous> (dist/tests/generator.test.js:72:29)

  2) testing
       generator-example (CLI)
         should run the default task using the `testing` name (global install):
     Error: Cannot find generator: "testing"
      at Generate.<anonymous> (node_modules/base-generators/index.js:499:18)
      at iterate (node_modules/async-each-series/index.js:8:5)
      at Function.module.exports (node_modules/async-each-series/index.js:20:3)
      at Generate.generate (node_modules/base-generators/index.js:492:15)
      at Generate.generate (node_modules/base-generators/index.js:463:23)
      at Context.<anonymous> (dist/tests/generator.test.js:89:11)

  3) testing
       generator-example (CLI)
         should run the default task using the `generator-example` generator alias (local generator.js):
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  

  4) testing
       generator-example (API)
         should run the `generator-example` task:
     Error: ENOENT: no such file or directory, stat '/tmp/testing/generate-generator-example/dist/tests/actual/example.txt'
  

  5) testing
       sub-generator
         should run the `generator:project` task when defined explicitly:
     Error: ENOENT: no such file or directory, stat '/tmp/testing/generate-generator-example/dist/tests/actual/example.txt'
  

  6) testing
       sub-generator
         should run tasks as a sub-generator:
     Error: ENOENT: no such file or directory, stat '/tmp/testing/generate-generator-example/dist/tests/actual/example.txt'
  



error Command failed with exit code 6.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

--- a/src/assets/templates/generate-swap-generator-main/example.txt
+++ b/src/assets/templates/generate-swap-generator-main/example.txt
@@ -3,6 +3,6 @@ layout: false
 rename:
   basename: 'src/assets/templates/example.txt'
 ---
-This is a example template for <%= ask('name') %> generator !
+This is a example template for <%= alias %> generator !
 
-By <%= ask('owner') %>, license <%= ask('license') %>
+By <%= owner %>, license <%= license %>

--- a/src/assets/templates/generate-swap-generator-main/generator-root.js
+++ b/src/assets/templates/generate-swap-generator-main/generator-root.js
@@ -3,4 +3,4 @@ layout: false
 rename:
   basename: 'generator.js'
 ---
-module.exports = require('./dist/lib/generator.js')
+module.exports = require('./dist/lib/generator.js').default

--- a/src/assets/templates/generate-swap-generator-main/generator-src.js
+++ b/src/assets/templates/generate-swap-generator-main/generator-src.js
@@ -15,7 +15,7 @@ import generateSubgeneratorExample from './subgenerators/generate-subgenerator-e
 import promptTask from './tasks/prompt'
 
 export default function (app) {
-  if (!isValid(app, '<%= ask('name') %>')) return
+  if (!isValid(app, '<%= packageName %>')) return
 
   app.on('error', err => app.log.error(err))
 

--- a/src/assets/templates/generate-swap-generator-main/generator.test.js
+++ b/src/assets/templates/generate-swap-generator-main/generator.test.js
@@ -69,7 +69,7 @@ describe('<%= packageName %>', function () {
 
     it('should run the default task using the `<%= alias %>` generator alias (local generator.js)', function (cb) {
       app.use(generator)
-      app.generate('project', exists('example.txt', cb))
+      app.generate('<%= alias %>', exists('example.txt', cb))
     })
   })
 
@@ -79,12 +79,12 @@ describe('<%= packageName %>', function () {
       app.generate('<%= alias %>', exists('example.txt', cb))
     })
 
-    it('should run the `<%= alias %>` task', function (cb) {
+    it('should run the `<%= alias %>:main` task', function (cb) {
       app.register('<%= alias %>', generator)
-      app.generate('<%= alias %>:project', exists('example.txt', cb))
+      app.generate('<%= alias %>:main', exists('example.txt', cb))
     })
 
-    it('should run the `default` task when defined explicitly', function (cb) {
+    it('should run the `default` task when called explicitly', function (cb) {
       app.register('<%= alias %>', generator)
       app.generate('<%= alias %>:default', exists('example.txt', cb))
     })
@@ -105,18 +105,18 @@ describe('<%= packageName %>', function () {
       app.generate('foo.<%= alias %>', exists('example.txt', cb))
     })
 
-    it('should run the `generator:default` task when defined explicitly', function (cb) {
+    it('should run the `<%= alias %>:default` task when defined explicitly', function (cb) {
       app.register('foo', function (foo) {
         foo.register('<%= alias %>', generator)
       })
       app.generate('foo.<%= alias %>:default', exists('example.txt', cb))
     })
 
-    it('should run the `generator:project` task when defined explicitly', function (cb) {
+    it('should run the `<%= alias %>:main` task when defined explicitly', function (cb) {
       app.register('foo', function (foo) {
         foo.register('<%= alias %>', generator)
       })
-      app.generate('foo.<%= alias %>:project', exists('example.txt', cb))
+      app.generate('foo.<%= alias %>:main', exists('example.txt', cb))
     })
 
     it('should work with nested sub-generators', function (cb) {
@@ -133,7 +133,7 @@ describe('<%= packageName %>', function () {
       const sub0 = app.register('sub0', generator)
       // eslint-disable-next-line no-unused-vars
       const sub1 = sub0.register('sub1', generator)
-      app.generate('sub0.sub1:package', exists('example.txt', cb))
+      app.generate('sub0.sub1:main', exists('example.txt', cb))
     })
   })
 

--- a/src/assets/templates/generate-swap-generator-main/generator.test.js
+++ b/src/assets/templates/generate-swap-generator-main/generator.test.js
@@ -22,7 +22,7 @@ const isTravis = isCI || process.env.TRAVIS
 // const fixtures = path.resolve.bind(path, __dirname, 'fixtures')
 const actual = path.resolve.bind(path, __dirname, 'actual')
 
-describe('<%= ask('name') %>', function () {
+describe('<%= packageName %>', function () {
   this.slow(250)
 
   if (isTravis) {
@@ -47,7 +47,7 @@ describe('<%= ask('name') %>', function () {
     it('should extend tasks onto the instance', function () {
       app.use(generator)
       expect(app.tasks).has.property('default')
-      expect(app.tasks).has.property('project')
+      expect(app.tasks).has.property('main')
     })
 
     it('should run the `default` task with .build', function (cb) {
@@ -62,9 +62,9 @@ describe('<%= ask('name') %>', function () {
   })
 
   describe('<%= alias %> (CLI)', function () {
-    it('should run the default task using the `<%= ask('name') %>` name (global install)', function (cb) {
+    it('should run the default task using the `<%= packageName %>` name (global install)', function (cb) {
       app.use(generator)
-      app.generate('<%= ask('name') %>', exists('example.txt', cb))
+      app.generate('<%= packageName %>', exists('example.txt', cb))
     })
 
     it('should run the default task using the `<%= alias %>` generator alias (local generator.js)', function (cb) {

--- a/src/assets/templates/generate-swap-generator-main/libs-subgenerator-example.js
+++ b/src/assets/templates/generate-swap-generator-main/libs-subgenerator-example.js
@@ -12,7 +12,7 @@ import { escapeQuotes } from '../../utils/helpers'
 import generateDefaults from 'generate-defaults'
 
 export default function (app) {
-  if (!isValid(app, 'generate-swap-generator-package')) return
+  if (!isValid(app, 'generate-subgenerator-example')) return
 
   app.on('error', err => app.log.error(err))
 

--- a/src/assets/templates/generate-swap-generator-main/plugin.test.js
+++ b/src/assets/templates/generate-swap-generator-main/plugin.test.js
@@ -13,7 +13,7 @@ let app
 
 const expect = chai.expect
 
-describe('<%= ask('name') %>', function () {
+describe('<%= packageName %>', function () {
   beforeEach(function () {
     app = generate()
   })
@@ -28,7 +28,7 @@ describe('<%= ask('name') %>', function () {
     it('should only register the plugin once', function (cb) {
       let count = 0
       app.on('plugin', function (name) {
-        if (name === '<%= ask('name') %>') {
+        if (name === '<%= packageName %>') {
           count++
         }
       })

--- a/src/assets/templates/generate-swap-generator-readme/readme.md
+++ b/src/assets/templates/generate-swap-generator-readme/readme.md
@@ -4,24 +4,24 @@ rename:
   basename: 'README.md'
 ---
 <h1 align="center">
-  <a href="https://github.com/sirap-group"><img src="src/assets/img/brand.png"/></a>
+  <a href="https://<%= defaultHost || ask('defaultHost') %>/<%= ask('namespace') %>"><img src="src/assets/img/brand.png"/></a>
   <br>
-  −− <%= ask('name') %> −−
+  −− <%= packageName || ask('packageName') %> −−
   <br>
   <br>
 </h1>
 
 <p align="center">
-  <a href="https://travis-ci.org/<%= ask('namespace') %>/<%= ask('name') %>" target="_blank">
-    <img alt="Build Status" src="https://travis-ci.org/<%= ask('namespace') %>/<%= ask('name') %>.svg?branch=master"/>
+  <a href="https://travis-ci.org/<%= ask('namespace') %>/<%= packageName || ask('packageName') %>" target="_blank">
+    <img alt="Build Status" src="https://travis-ci.org/<%= ask('namespace') %>/<%= packageName || ask('packageName') %>.svg?branch=master"/>
   </a>
 
-  <a href="https://www.npmjs.com/package/<%= ask('name') %>">
-    <img src="https://img.shields.io/npm/dm/<%= ask('name') %>.svg" alt="npm downloads">
+  <a href="https://www.npmjs.com/package/<%= packageName || ask('packageName') %>">
+    <img src="https://img.shields.io/npm/dm/<%= packageName || ask('packageName') %>.svg" alt="npm downloads">
   </a>
 
-  <a href="https://npmjs.org/package/<%= ask('name') %>" target="_blank">
-    <img alt="NPM version" src="https://badge.fury.io/js/<%= ask('name') %>.svg"/>
+  <a href="https://npmjs.org/package/<%= packageName || ask('packageName') %>" target="_blank">
+    <img alt="NPM version" src="https://badge.fury.io/js/<%= packageName || ask('packageName') %>.svg"/>
   </a>
 
   <a href="https://standardjs.com" target="_blank">
@@ -33,8 +33,8 @@ rename:
   −− <%= ask('description') %> −−
   <br>
   <br>
-  <a href="https://www.npmjs.com/package/<%= ask('name') %>" target="_blank">
-    <img alt="NPM Stats" src="https://nodei.co/npm/<%= ask('name') %>.png?downloads=true&downloadRank=true&stars=true"/>
+  <a href="https://www.npmjs.com/package/<%= packageName || ask('packageName') %>" target="_blank">
+    <img alt="NPM Stats" src="https://nodei.co/npm/<%= packageName || ask('packageName') %>.png?downloads=true&downloadRank=true&stars=true"/>
   </a>
 </h4>
 
@@ -43,13 +43,13 @@ rename:
 ### Global install to use the generator as a standarlone CLI (recommanded)
 
 ```sh
-$ yarn gobal add <%= ask('name') %>
+$ yarn gobal add <%= packageName || ask('packageName') %>
 ```
 
 or
 
 ```sh
-$ npm install --global <%= ask('name') %>
+$ npm install --global <%= packageName || ask('packageName') %>
 ```
 
 ### Local install to use as a `generate` plugin or sub-generator
@@ -57,13 +57,13 @@ $ npm install --global <%= ask('name') %>
 > You should first visit the generator framework, [generate](https://github.com/generate/generate) to understand how generators, subgenerators and plugins work. See also [related projects](#related-projects)
 
 ```sh
-$ yarn add <%= ask('name') %>
+$ yarn add <%= packageName || ask('packageName') %>
 ```
 
 or
 
 ```sh
-$ npm install --save <%= ask('name') %>
+$ npm install --save <%= packageName || ask('packageName') %>
 ```
 
 ## Usage
@@ -71,7 +71,7 @@ $ npm install --save <%= ask('name') %>
 ### As a standalone generator (recommanded)
 
 ```sh
-$ <%= ask('name') %>
+$ <%= packageName || ask('packageName') %>
 ```
 
 Should **@TODO: describe generator default action**
@@ -101,50 +101,50 @@ Should **@TODO: describe generator default action**
 > You should first visit the generator framework, [generate](https://github.com/generate/generate) to understand how generators, subgenerators and plugins work.
 
 ```js
-import <%= camelcase(ask('name')) %> from '<%= ask('name') %>'
+import <%= camelcase(packageName || ask('packageName')) %> from '<%= packageName || ask('packageName') %>'
 import isValid from 'is-valid-app'
 
 export default function (app) {
   if (!isValid(app, 'generate-my-generator')) return
 
-  app.use(<%= camelcase(ask('name')) %>)
+  app.use(<%= camelcase(packageName || ask('packageName')) %>)
 
   // ... your generator's stuff
 }
 ```
 
-Then you can use directly any '<%= ask('name') %>' tasks as a `generate-my-generator` task.
+Then you can use directly any '<%= packageName || ask('packageName') %>' tasks as a `generate-my-generator` task.
 
 ```
 $ gen my-generator:example-task
 ```
 
-Should run `<%= ask('name') %>:example-task` on your own generator instance.
+Should run `<%= packageName || ask('packageName') %>:example-task` on your own generator instance.
 
 ### As a generate sub-generator (if you know what you are doing)
 
 > You should first visit the generator framework, [generate](https://github.com/generate/generate) to understand how generators, subgenerators and plugins work.
 
 ```js
-import <%= camelcase(ask('name')) %> from '<%= ask('name') %>'
+import <%= camelcase(packageName || ask('packageName')) %> from '<%= packageName || ask('packageName') %>'
 import isValid from 'is-valid-app'
 
 export default function (app) {
   if (!isValid(app, 'generate-my-generator')) return
 
-  app.register('<%= ask('alias') %>', <%= camelcase(ask('name')) %>)
+  app.register('<%= ask('alias') %>', <%= camelcase(packageName || ask('packageName')) %>)
 
   // ... your generator's stuff
 }
 ```
 
-Then you can acces easily any `<%= ask('name') %>` tasks as a `generate-my-generator` sub-generator task.
+Then you can acces easily any `<%= packageName || ask('packageName') %>` tasks as a `generate-my-generator` sub-generator task.
 
 ```
 $ gen my-generator.<%= ask('alias') %>:example-task
 ```
 
-Should run `<%= ask('name') %>:example-task` on your own generator instance.
+Should run `<%= packageName || ask('packageName') %>:example-task` on your own generator instance.
 
 ## API
 
@@ -196,8 +196,8 @@ All the rest of the development dependencies are local.
 Clone the repo and install dependencies:
 
 ```sh
-$ git clone git@github.com:<%= ask('namespace') %>/<%= ask('name') %>.git
-$ cd <%= ask('name') %>
+$ git clone git@github.com:<%= ask('namespace') %>/<%= packageName || ask('packageName') %>.git
+$ cd <%= packageName || ask('packageName') %>
 $ yarn install
 ```
 #### Running test

--- a/src/lib/subgenerators/generate-swap-generator-main/generator.js
+++ b/src/lib/subgenerators/generate-swap-generator-main/generator.js
@@ -28,7 +28,7 @@ export default function (app) {
    * @name main
    * @api public
    */
-  app.task('main', ['set-default-githost', 'index-root', 'generator-root', 'generator-src', 'generator-test', 'plugin-test', 'template-example', 'libs'])
+  app.task('main', ['set-default-githost', 'index-root', 'generator-root', 'generator-src', 'generator-test', 'plugin-test', 'templates', 'libs'])
 
   app.task('set-default-githost', done => {
     const githosts = app.base.data('githosts')
@@ -116,7 +116,8 @@ export default function (app) {
     .on('finish', done)
     .on('error', done)
   })
-
+  task(app, 'template-example-subgenerator', 'generate-swap-generator-main/example-subgenerator.txt')
+  app.task('templates', ['template-example', 'template-example-subgenerator'])
   /**
    * Copy the lib files
    *

--- a/src/lib/tasks/prompt.js
+++ b/src/lib/tasks/prompt.js
@@ -36,9 +36,15 @@ export default app => {
     askPromise(['alias'])
     .then(({alias}) => {
       name = `generate-${alias}`
-      !app.option('silent') && app.log.success(`Required package name is "${name}" (related to the generator alias)`)
-      app.base.data({alias, name})
 
+      !app.option('silent') && app.log.success(`Required package name is "${name}" (related to the generator alias)`)
+
+      app.base.data({alias, name, packageName: name})
+
+      app.question('name', {
+        message: 'Package name (You should keep the suggestion or the generator may be broken) ?',
+        default: name
+      })
       app.question('dest', {
         message: 'Project directory ?',
         default: name
@@ -58,7 +64,7 @@ export default app => {
         message: 'Author name ?'
       })
 
-      return askPromise(['dest', 'description', 'githosts', 'author.username', 'author.name'])
+      return askPromise(['name', 'dest', 'description', 'githosts', 'author.username', 'author.name'])
     })
     .then(answers => {
       const {githosts} = answers

--- a/src/lib/utils/utils.js
+++ b/src/lib/utils/utils.js
@@ -14,6 +14,9 @@ function task (app, name, pattern, dependencies) {
 function file (app, pattern) {
   const opts = extend({}, app.base.options, app.options)
   const srcBase = opts.srcBase || path.join(__dirname, '../../assets/templates')
+
+  app.base.set('cache.data.name', app.base.cache.data.packageName)
+
   return app.src(pattern, {cwd: srcBase})
     .pipe(app.renderFile('*', app.base.cache.data))
     .pipe(app.conflicts(app.cwd))

--- a/src/tests/generator.test.js
+++ b/src/tests/generator.test.js
@@ -145,6 +145,7 @@ describe('generate-swap-project', function () {
 
     app.data(pkg)
     app.data('project', pkg)
+    app.data('packageName', pkg.name)
     app.data('username', 'foo')
     app.data('namespace', 'myorg')
     app.data('owner', 'foo')

--- a/src/tests/generator.test.js
+++ b/src/tests/generator.test.js
@@ -64,7 +64,7 @@ describe('generate-swap-project', function () {
 
     it('should run the default task using the `swap-project` generator alias (local generator.js)', function (cb) {
       app.use(generator)
-      app.generate('project', exists('package.json', cb))
+      app.generate('swap-project', exists('package.json', cb))
     })
   })
 

